pkg/archive/filter.go

```diff
@@ -0,0 +1,78 @@
+package archive
+
+import (
+	"bufio"
+	"io"
+	"os/exec"
+	"sync"
+
+	"github.com/containers/storage/pkg/pools"
+)
+
+var (
+	procFilterLock sync.Mutex
+	filterPath     map[string]string
+)
+
+type filterReader struct {
+	reader *io.PipeReader
+	runErr error
+}
+
+func (f *filterReader) Read(p []byte) (n int, err error) {
+	if f.runErr != nil {
+		return 0, f.runErr
+	}
+	return f.reader.Read(p)
+}
+
+func (f *filterReader) Close() error {
+	err := f.reader.Close()
+	if err == nil {
+		err = f.runErr
+	}
+	return err
+}
+
+func getFilterPath(name string) string {
+	procFilterLock.Lock()
+	if filterPath == nil {
+		filterPath = make(map[string]string)
+	}
+	path, ok := filterPath[name]
+	procFilterLock.Unlock()
+
+	if ok {
+		return path
+	}
+
+	path, err := exec.LookPath(name)
+	if err != nil {
+		path = ""
+	}
+
+	procFilterLock.Lock()
+	filterPath[name] = path
+	procFilterLock.Unlock()
+	return path
+}
+
+func tryProcFilter(args []string, pool *pools.BufioReaderPool, input *bufio.Reader) (io.ReadCloser, bool) {
+	path := getFilterPath(args[0])
+	if path == "" {
+		return nil, false
+	}
+
+	r, w := io.Pipe()
+	cmd := exec.Command(path, args[1:]...)
+	cmd.Stdin = input
+	cmd.Stdout = w
+
+	f := &filterReader{reader: r}
+	go func() {
+		defer w.Close()
+		f.runErr = cmd.Run()
+		pool.Put(input)
```

Comment:
I don’t think this function should be in the business of managing `pool`; let the caller worry about that, probably using `NewReadCloserWrapper` like on all the other paths.

(Uh… is the existing code in the caller safe? Shouldn’t it trigger on close of the _input_ to the decompression, rather than the _output_? It’s not obvious to me that they always happen in the expected order.

If it is not safe, that’s _also_ a reason for the caller to worry about that … and at least a separate commit.)